adds [*mut|*const] ptr::set_ptr_value

oliver-giersch opened the pull request:
I propose the addition of these two functions to `*mut T` and `*const T`, respectively. The motivation for this is primarily byte-wise pointer arithmetic on (potentially) fat pointers, i.e. for types with a `T: ?Sized` bound. A concrete use-case has been discussed in [this](https://internals.rust-lang.org/t/byte-wise-fat-pointer-arithmetic/12739) thread.
TL;DR: Currently, byte-wise pointer arithmetic with potentially fat pointers in not possible in either stable or nightly Rust without making assumptions about the layout of fat pointers, which is currently still an implementation detail and not formally stabilized. This PR adds one function to `*mut T` and `*const T` each, allowing to circumvent this restriction without exposing any internal implementation details.
One possible alternative would be to add specific byte-wise pointer arithmetic functions to the two pointer types in addition to the already existing count-wise functions. However, I feel this fairly niche use case does not warrant adding a whole set of new functions like `add_bytes`, `offset_bytes`, `wrapping_offset_bytes`, etc. (times two, one for each pointer type) to `libcore`.

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

oliver-giersch commented:
Done. Although I always thought documentation was generated alphabetically, so in-file function order did not matter?

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

oliver-giersch commented:
made a complete mess trying to merge the recent directory structure changes, but everything should be ok again

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 6c81556a36ac5507fe1f9cd8ee699e6fa2b11077 has been approved by `dtolnay`

<!-- @bors r=dtolnay 6c81556a36ac5507fe1f9cd8ee699e6fa2b11077 -->
<!-- homu: {"type":"Approved","sha":"6c81556a36ac5507fe1f9cd8ee699e6fa2b11077","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 6c81556a36ac5507fe1f9cd8ee699e6fa2b11077 with merge 24265c7b79dcb81fb22b9d4c8e14acb89cb7a736...
<!-- homu: {"type":"BuildStarted","head_sha":"6c81556a36ac5507fe1f9cd8ee699e6fa2b11077","merge_sha":"24265c7b79dcb81fb22b9d4c8e14acb89cb7a736"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

oliver-giersch commented:
What's the significance of the test time out and what can I do about it?

ecstatic-morse commented:
@bors retry

RalfJung commented:
I think there is some code in `Rc`/`Arc` that could benefit from this... Cc @CAD97

HeroicKatora commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -656,6 +656,38 @@ impl<T: ?Sized> *const T {
+    ///
+    /// In case `self` is a (fat) pointer to an unsized type, this operation
+    /// will only affect the pointer part, whereas for (thin) pointers to
+    /// sized types, this has the same effect as a simple assignment.
```
This should discuss pointer provenance. Will the resulting pointer have provenance of `self` or of `val`? I think it will be quite a bit less useful to have that of `self` as the usecase of pivoting a pointer to a copy of the bytes (i.e. 'moving' an unsized value) couldn't work otherwise.

CAD97 replied:
I agree on both counts: provenance needs to be mentioned in some fashion by the docs, and those semantics (takes provenance from val, not self) are what makes sense.

Basically, fat_ptr.set_ptr_value(new_ptr) should have the semantics of create_fat_pointer(new_ptr, ptr_metadata_of(fat_ptr)).

I don't think there's a way in which "physical bits of val but provenance of self" makes sense, but it's still good to be clear that provenance of the new value is kept alongside its target.

RalfJung replied:
The provenance will be that of `val`. There is currently no operation in Rust that can take the provenance of one pointer and combine it with the value (the memory location it points to) of another pointer.

CAD97 commented:
Mild concern: this is `fn(*mut T, *mut ()) -> *mut T`, but with the name `set_ptr_value` it could be assumed to mutate in place instead of create a new pointer. At the very minimum these functions should be marked `#[must_use]` to help prevent that mistake.

[The version of this in the `A`/`Rc` implementation](https://github.com/rust-lang/rust/blob/71f8d0c8f1060bbe74100f29cc6f2da63d697c28/library/alloc/src/rc.rs#L989-L998) also does the same thing (`fn set_data_ptr<T: ?Sized, U>(mut ptr: *mut T, data: *mut U) -> *mut T`), but it also doesn't use method syntax, so it's clear that it returns a new value at the use site (as no `&mut` is provided), unlike with method syntax.

CAD97 commented:
Mild concern: a main usecase of this is to implement the [`wrapping_`]`offset` operation for potentially fat pointers, with a byte offset instead of an element-wise offset. I still think it's worth doing this; it's not _that_ uncommon for optimized raw-pointer-heavy data structures to store a pointer not to the whole allocated chunk but to the "interesting" part of the allocation.

Offsetting a pointer to sized type by a byte offset is `(ptr as *mut u8).offset(offset) as *mut T`, offsetting a pointer to a potentially unsized type by a byte offset with this PR would be `(ptr as *mut T).set_ptr_value((ptr as *mut u8).offset(offset) as *mut ())`.

At the least, I think it'd be better for the "pointer to data" pointer for these methods to be `*mut u8` instead of `*mut ()` so it's a lot harder to write the shorter and _incorrect_ `(ptr as *mut T).set_ptr_value((ptr as *mut ()).offset(offset))`. Clippy catches calling `offset` on a pointer to a ZST now (after I made the mistake and had a nasty bug in a refactor due to it (thankfully a miri-caught one)), but it'd be nice to avoid making this mistake easier to write.

oliver-giersch commented:
There are two identical definitions of a similar `set_data_ptr` function [here](https://github.com/rust-lang/rust/blob/63e34422bbaf4ae4ed5ae7309183185aa2aa13a4/library/alloc/src/sync.rs#L948) and [here](https://github.com/rust-lang/rust/blob/63e34422bbaf4ae4ed5ae7309183185aa2aa13a4/library/alloc/src/rc.rs#L993) with four separate usages each.

In the last commit I've added `#[must_use]` attributes, mentioned provenance in the documentation and changed the type of `val` to `*mut u8`.

oliver-giersch commented:
Ok, I am a bit confused by the merge process. So the initial changes have been merged as part of a roll-up, but not that latest commit I've made addressing the latest change requests and neither has the PR itself been merged.
Are there any actions I should be taking in all this or do these things just take time until someone with the right permissions comes about?

RalfJung commented:
@oliver-giersch looks like you pushed something after [approval](https://github.com/rust-lang/rust/pull/74774#issuecomment-667966098) -- in fact even after rollup, so it was not considered.

Can you make a new PR with just those remaining changes?

oliver-giersch commented:
Got it, I wasn't aware that I wasn't supposed to make changes after approval, will keep that in mind for further contributions. The rollup also occurred only a few minutes before I made the final commit. Should I close this PR now?

RalfJung commented:
Yeah it's better to wait for someone to `bors r-` before pushing again. Looks like it was a close race.^^

Yes please close this PR and make a new one for what got missed. Great that you noticed this. :)
